internal/command/deploy/machines.go

```diff
@@ -207,14 +211,129 @@ func (md *machineDeployment) runReleaseCommand(ctx context.Context) error {
 	return nil
 }
 
+func (md *machineDeployment) resolveProcessGroupChanges() ProcessGroupsDiff {
+
+	output := ProcessGroupsDiff{
+		groupsToRemove:        map[string]int{},
+		groupsNeedingMachines: map[string]*appconfig.ProcessConfig{},
+	}
+
+	groupHasMachine := map[string]bool{}
+
+	for _, leasableMachine := range md.machineSet.GetMachines() {
+		mach := leasableMachine.Machine()
+		machGroup := mach.ProcessGroup()
+		groupMatch := ""
+		for name := range md.processConfigs {
+			if machGroup == name {
+				groupMatch = machGroup
+				break
+			}
+		}
+		if groupMatch == "" {
+			output.groupsToRemove[groupMatch] += 1
+			output.machinesToRemove = append(output.machinesToRemove, leasableMachine)
+		} else {
+			groupHasMachine[groupMatch] = true
+		}
+	}
+
+	for name, val := range md.processConfigs {
+		if hasMach, ok := groupHasMachine[name]; !ok || !hasMach {
+			output.groupsNeedingMachines[name] = val
+		}
+	}
+
+	return output
+}
+
+func (md *machineDeployment) warnAboutProcessGroupChanges(ctx context.Context, diff ProcessGroupsDiff) {
+
+	var (
+		io                 = iostreams.FromContext(ctx)
+		colorize           = io.ColorScheme()
+		willAddMachines    = len(diff.groupsNeedingMachines) != 0
+		willRemoveMachines = diff.machinesToRemove != nil
+	)
+
+	if willAddMachines || willRemoveMachines {
+		fmt.Fprintln(io.ErrOut, "Process groups have changed. This will:")
+	}
+	// TODO(ali): Figure out why the bullets aren't colored
+	if willRemoveMachines {
+		bullet := colorize.Red("*")
+		for grp, numMach := range diff.groupsToRemove {
+			pluralS := lo.Ternary(numMach == 1, "", "s")
+			fmt.Fprintf(io.ErrOut, " %s destroy %d \"%s\" machine%s\n", bullet, numMach, grp, pluralS)
+		}
+	}
+	if willAddMachines {
+		bullet := colorize.Green("*")
+
+		for name := range diff.groupsNeedingMachines {
+			fmt.Fprintf(io.ErrOut, " %s create 1 \"%s\" machine\n", bullet, name)
+		}
+	}
+
+}
+
+func (md *machineDeployment) spawnMachineInGroup(ctx context.Context, groupName string) error {
+	fmt.Fprintf(md.io.Out, "No machines in group '%s', launching one new machine\n", md.colorize.Bold(groupName))
+	machBase := &api.Machine{
+		Region: md.appConfig.PrimaryRegion,
+		Config: &api.MachineConfig{
+			Metadata: map[string]string{
+				api.MachineConfigMetadataKeyFlyProcessGroup: groupName,
+			},
+		},
+	}
+	launchInput := md.resolveUpdatedMachineConfig(machBase, false)
+	newMachineRaw, err := md.flapsClient.Launch(ctx, *launchInput)
+	newMachine := machine.NewLeasableMachine(md.flapsClient, md.io, newMachineRaw)
+	if err != nil {
+		return fmt.Errorf("error creating a new machine machine: %w", err)
```

Comment:
> The deletion hook, I guess?

Yeah. I saw that and figured it was probably important to make sure behavior stayed consistent no matter how a machine was deleted. Right now it's just one hook, but I don't want to introduce a gap where the behavior can slip further apart over time.

A lot of commands export public helpers. It's clunky, but it helps to work around go's limitation against reference cycles.

> I think the deploy code already has access to an *api.Machine to pass in, so that would avoid pulling down another copy for each machine it's going to destroy

Good call

> it would essentially remove a conflict between these changes and https://github.com/superfly/flyctl/pull/1899.

I'll wait for that pr to land, then update this to do exactly that :)